### PR TITLE
Adding gpsd support in collectd.

### PR DESCRIPTION
**Pull request by NicolasJourden**

### CODE REVIEW

**Comment by octo** [comment]
Hi @NicolasJourden, thank you very much for your patch! Can you please chose a license for your code and add it to the beginning of the `gps.c` file? For new code, we're preferring the MIT license, which you can find in the `COPYING` file. Take a look at, for example, `src/mqtt.c` and copy its header to your file.

I'll leave some code review comments inline.

Thanks again and best regards,
—octo

File: src/gps.pod [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+The host on which gpsd runs.
+
+=item B<Port>
+
+Port to connect to gpsd on the host machine.
+
+=item B<Timeout>
+
+Timeout in ms.
```
Comment (octo) [inline src/gps.pod]: We're trying to default to seconds, rather than having a mix of seconds, milliseconds, microseconds, … in config options. Ideally, you'd use the type `cdtime_t` to store the value in your code.

File: src/gps.c [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+/**
+ * Read the data and submit.
+ */
+static int gps_collectd_read ()
+{
+  pthread_mutex_lock (&data_mutex);
+  gps_collectd_submit("gps_hdop", (gauge_t) gps_data_read.hdop);
+  gps_collectd_submit("gps_vdop", (gauge_t) gps_data_read.vdop);
+  gps_collectd_submit("gps_sat", (gauge_t) gps_data_read.satellites);
+  printf ("gps: hdop=%1.3f, vdop=%1.3f, sat=%02d.\n",
```
Comment (octo) [inline src/gps.c]: Please use the `DEBUG()` macro rather than `printf()` for such messages.

File: src/gps.c [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+
+/**
+ * Read configuration.
+ */
+static int gps_collectd_config (const char *key, const char *value)
+{
+  if (strcasecmp (key, "Host") == 0) {
+    if (gps_data_config.host != NULL) free (gps_data_config.host);
```
Comment (octo) [inline src/gps.c]: This indentation is extremely misleading. Please add the `free()` on a line of its own and make sure the `gpd_data_config` assignment is on the same indentation as the `if`.

File: src/gps.c [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      gps_data_config.host = sstrdup (value);
+  }
+  if (strcasecmp (key, "Port") == 0) {
+    if (gps_data_config.port != NULL) free (gps_data_config.port);
```
Comment (octo) [inline src/gps.c]: Same as above: fix indentation

File: src/gps.c [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+
+/**
+ * Init.
+ */
+static int gps_collectd_init (void)
+{
+  int err = 0;
+
+  printf ("gps: will use %s:%s with timeout %d.\n", gps_data_config.host, gps_data_config.port, gps_data_config.timeout);
```
Comment (octo) [inline src/gps.c]: `DEBUG(…)`

File: src/gps.c [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+  err = plugin_thread_create (&connector, NULL, gps_collectd_thread, NULL);
+
+  if (err != 0) {
+    WARNING ("pthread_create() failed.");
```
Comment (octo) [inline src/gps.c]: This should be an `ERROR()`, since you can't recover from this.

File: src/gps.c [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+  return (0);
+}
+
+/**
+ * Register the module.
+ */
+void module_register (void)                                                                    
+{
+  gps_data_config.host = sstrdup ("localhost");
```
Comment (octo) [inline src/gps.c]: Please add the default values into a define, i.e. something like this near the top of the file:
```c
#define GPS_DEFAULT_NODE "localhost"
#define GPS_DEFAULT_SERVICE "2947"
```

File: src/gps.c [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+  int timeout;
+} gps_definition_t;
+static gps_definition_t gps_data_config;
+
+typedef struct {
+ int satellites;
+ double vdop;
+ double hdop;
+} gpsDATA_t;
+static gpsDATA_t gps_data_read;
```
Comment (octo) [inline src/gps.c]: Please don't use uppsercase characters in type names.

File: src/gps.c [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            gps_data_read.satellites = gps_data.satellites_used;
+          }
+
+          pthread_mutex_unlock (&data_mutex);
+        }
+      }
+    }
+  }
+
+  gps_stream(&gps_data, WATCH_DISABLE, NULL);
```
Comment (octo) [inline src/gps.c]: I don't think this code is ever reached.

File: src/types.db [hunk 10]
```diff
@@ -78,6 +78,9 @@ frequency_offset	value:GAUGE:-1000000:1000000
 frequency		value:GAUGE:0:U
 fscache_stat		value:DERIVE:0:U
 gauge			value:GAUGE:U:U
+gps_sat			value:GAUGE:0:25
+gps_hdop		value:GAUGE:0:100
```
Comment (octo) [inline src/types.db]: Can we please rename `gps_hdop` and `gps_vdop` to `dilution_of_precision` and set the *type instance* to `horizontal` and `vertical`?

File: src/types.db [hunk 11]
```diff
@@ -78,6 +78,9 @@ frequency_offset	value:GAUGE:-1000000:1000000
 frequency		value:GAUGE:0:U
 fscache_stat		value:DERIVE:0:U
 gauge			value:GAUGE:U:U
+gps_sat			value:GAUGE:0:25
```
Comment (octo) [inline src/types.db]: Please rename this type to `satellites` and remove the upper limit (i.e. set it to `U`).

**Comment by mfournier** [comment]
FYI I updated the jenkins setup to install libgps header files on debian/ubuntu as well as EL7. So once you update the pull-request, these test environments won't skip building this plugin.

**Comment by NicolasJourden** [comment]
Hi,

I applied the requested modification, let me know if that works for you.
Only missing item is the cdtime_t.

Cheers,
Nicolas

**Comment by octo** [comment]
Thanks for the update @NicolasJourden! I've done some coding style fixes and implemented the `cdtime_t` storage of the timeout and pause options. Can you please merge [the pr/1346 branch](https://github.com/octo/collectd/tree/pr/1346) into this PR and give it a go?

```sh
git fetch git://github.com/octo/collectd.git pr/1346
git merge FETCH_HEAD
```

Since you mention NTP in the documentation: does it make sense to store the time error which libgps presumably provides too?

**Comment by NicolasJourden** [comment]
Hey @octo o, that should be okay now.
I also added some features to detect if gpsd service goes down and reconnect... better than restarting collectd service.
Also it requires ms not us as you wrote.
I tested it with a gps and gpsd, it works fine.

**Comment by octo** [comment]
tl;dr: LGTM, but see comments.

Hi @NicolasJourden, thank you for the update. The code looks good and ready to merge, with exception of the ms / us confusion:

I'm quite certain that _libgps_ expects microseconds, though: [libgps(3)](http://www.catb.org/gpsd/libgps.html) talks about "microseconds" and both, [libgps_sock.c](http://git.savannah.gnu.org/cgit/gpsd.git/tree/libgps_sock.c#n106) and [libgps_shm.c](http://git.savannah.gnu.org/cgit/gpsd.git/tree/libgps_shm.c#n79) both expect to get µs by multiplying / dividing by 10e6.

Can you please give the code with us instead of ms a try? I'd the number of timeouts reported by `gps_waiting()` to drop to zero with the us code.

Thanks and best regards,
—octo

**Comment by mfournier** [comment]
@NicolasJourden the plugin fails to build against older version of libgps (noticed with 2.95). Fortunately it seems pretty easy to make it compatible. Can you take a look at https://github.com/mfournier/collectd/commit/5579b150 and pull it in if you agree ?

I also merged `gps.pod` into `collectd.conf.pod`. I agree the collectd.conf manpage is huge, but let's adress this independently. Can you merge https://github.com/mfournier/collectd/commit/8884f852e in please ?

Finally, it seems there's a small memory leak. Any idea ?
```
==13145== HEAP SUMMARY:
==13145==     in use at exit: 125,109 bytes in 125 blocks
==13145==   total heap usage: 1,704 allocs, 1,579 frees, 385,697 bytes allocated
==13145== 
==13145== 320 bytes in 1 blocks are possibly lost in loss record 30 of 39
==13145==    at 0x4C2AC15: calloc (vg_replace_malloc.c:711)
==13145==    by 0x4010FE1: allocate_dtv (dl-tls.c:296)
==13145==    by 0x40116ED: _dl_allocate_tls (dl-tls.c:460)
==13145==    by 0x513EC27: allocate_stack (allocatestack.c:589)
==13145==    by 0x513EC27: pthread_create@@GLIBC_2.2.5 (pthread_create.c:495)
==13145==    by 0x6310598: cgps_init (gps.c:235)
==13145==    by 0x40D749: plugin_init_all (plugin.c:1744)
==13145==    by 0x4062AB: do_init (collectd.c:329)
==13145==    by 0x4062AB: main (collectd.c:719)
==13145== 
==13145== LEAK SUMMARY:
==13145==    definitely lost: 0 bytes in 0 blocks
==13145==    indirectly lost: 0 bytes in 0 blocks
==13145==      possibly lost: 320 bytes in 1 blocks
==13145==    still reachable: 124,789 bytes in 124 blocks
==13145==         suppressed: 0 bytes in 0 blocks
==13145== Reachable blocks (those to which a pointer was found) are not shown.
==13145== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==13145== 
==13145== For counts of detected and suppressed errors, rerun with: -v
==13145== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

Many thanks !

**Comment by NicolasJourden** [comment]
Hello,

I accepted the changes @octo & @mfournier. I fixed some issues this weekend and tested again.
My bad for the units ;-)
@octo , I don't understand your idea to drop to zero, you want to count down errors?

Thanks.

**Comment by mfournier** [comment]
Thanks @NicolasJourden !

Unfortunately valgrind still seems to detect mistakes with the current code. Can you try running `valgrind --leak-check=full /path/to/sbin/collectd -f` to see by yourself ?

I also noticed a `Conditional jump or move depends on uninitialised value(s)` warning in valgrind. Not sure is was already there before the last updates or not though...

Thanks !

**Comment by NicolasJourden** [comment]
@mfournier , I did some changes in the way of closing the reading thread and get something like that:

==24751== Memcheck, a memory error detector
==24751== Copyright (C) 2002-2011, and GNU GPL'd, by Julian Seward et al.
==24751== Using Valgrind-3.7.0 and LibVEX; rerun with -h for copyright info
==24751== Command: src/daemon/collectd -f
==24751== 
option = Hostname; value = testpc;
option = BaseDir; value = /opt/collectd/var/lib/collectd;
Done parsing `/opt/collectd/share/collectd/types.db'
option = Interval; value = 10.000000;
Created new plugin context.
^C==24751== 
==24751== HEAP SUMMARY:
==24751==     in use at exit: 90,442 bytes in 243 blocks
==24751==   total heap usage: 2,061 allocs, 1,818 frees, 990,769 bytes allocated
==24751== 
==24751== LEAK SUMMARY:
==24751==    definitely lost: 0 bytes in 0 blocks
==24751==    indirectly lost: 0 bytes in 0 blocks
==24751==      possibly lost: 0 bytes in 0 blocks
==24751==    still reachable: 90,442 bytes in 243 blocks
==24751==         suppressed: 0 bytes in 0 blocks
==24751== Reachable blocks (those to which a pointer was found) are not shown.
==24751== To see them, rerun with: --leak-check=full --show-reachable=yes
==24751== 
==24751== For counts of detected and suppressed errors, rerun with: -v
==24751== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 6 from 6)

I will do more tests this week, hope we can finally merge soon :-)

File: src/gps.c [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+  return (0);
+}
+
+/**
+ * Shutdown.
+ */
+static int cgps_shutdown (void)
+{
+  if (connector != ((pthread_t) 0))
```
Comment (octo) [inline src/gps.c]: This is not portable: on some systems `pthread_t` is a struct. We *could* do something along the lines of:

```c
pthread_t nill;
memset (nill, 0, sizeof (nill));
if (memcmp (&connector, &nill, sizeof (nill)) != 0) {
  // …
}
```

But I would argue that introducing a `_Bool connector_running` variable to track this is *much* easier to read.

File: src/gps.c [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+      WARNING ("gps plugin: Ignoring unknown config option \"%s\".", child->key);
+  }
+
+  return 0;
+}
+
+/**
+ * Init.
+ */
+static int cgps_init (void)
```
Comment (octo) [inline src/gps.c]: The *init* callback may be called multiple times. I'd propose to use the `connector_running` boolean mentioned below to protect against multiple calls of this callback:

```c
if (connector_running)
  return 0;
```

**Comment by octo** [comment]
Thanks for all your work, @NicolasJourden!

**Comment by NicolasJourden** [comment]
@octo, changed the code... it should be better now.
@mfournier, the memory issue was in the way the thread was killed,  changed the way to do it. On my side there is not more issue, can you try on yours?
Let me know if there are stuff you would like to see changed.
I tested it today again with a GPS, looks fine.

**Comment by mfournier** [comment]
Looks great from a valgrind user point of view now :-) I tested with and without a gps connected to gpsd, and with gpsd stopped. Thanks for the update @NicolasJourden !

I'll leave a couple of comments inline for a few details I spotted.

File: src/collectd.conf.pod [hunk 14]
```diff
@@ -2482,6 +2482,74 @@ source, this is optional. Otherwise the option is required.
...
+
+Port to connect to gpsd on the host machine. Defaults to B<2947>.
+
+=item B<Timeout> I<Seconds>
+
+Timeout in seconds (default 0.015 sec).
+
+=item B<PauseRead> I<Seconds>
+
+Pause to apply between readings in seconds (default 1 sec).
```
Comment (mfournier) [inline src/collectd.conf.pod]: The default value mentioned here is not consistent with what is in the synopsis above.

File: src/collectd.conf.pod [hunk 15]
```diff
@@ -2482,6 +2482,74 @@ source, this is optional. Otherwise the option is required.
...
+
+=item B<PauseConnect> I<Seconds>
+
+Pause to apply between tentative of connection in seconds (default 1 sec).
```
Comment (mfournier) [inline src/collectd.conf.pod]: When reading the description of these 3 config options, I don't really understand what they refer to. Maybe expanding a little could help, or a few words in the introduction explaining how the gps device is probed by the gpsd daemon, and how collectd connects to gpsd ?

File: src/collectd.conf.pod [hunk 16]
```diff
@@ -2482,6 +2482,74 @@ source, this is optional. Otherwise the option is required.
...
+ LoadPlugin gps
+ <Plugin "gps">
+   # Connect to localhost on gpsd regular port:
+   Host "127.0.0.1"
+   Port "2947"
+   # 15 ms timeout
+   Timeout 0.015
+   # PauseRead of 5 sec. between readings:
+   PauseRead 5
+   # PauseConnect of 1 sec. between connection tentatives:
```
Comment (mfournier) [inline src/collectd.conf.pod]: "connection attempts"

File: src/collectd.conf.pod (hunk 15 above)
Comment (mfournier) [inline src/collectd.conf.pod]: "connection attempts" here too

File: src/gps.c [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+    int status = gps_open (cgps_config_data.host, cgps_config_data.port, &gpsd_conn);
+#else
+    int status = gps_open_r (cgps_config_data.host, cgps_config_data.port, &gpsd_conn);
+#endif
+    if (status < 0)
+    {
+      WARNING ("gps plugin: connecting to %s:%s failed: %s",
+               cgps_config_data.host, cgps_config_data.port, gps_errstr (status));
+
+      // Here we make a pause until a new tentative to connect, we check also if
```
Comment (mfournier) [inline src/gps.c]: ditto, although users won't notice this one ;)

File: src/gps.c [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+#include "common.h"
+#include "plugin.h"
+#include "utils_time.h"
+#include "configfile.h"
+
+#define CGPS_TRUE                  1
+#define CGPS_FALSE                 0
+#define CGPS_DEFAULT_HOST          "localhost"
+#define CGPS_DEFAULT_PORT          "2947"
+#define CGPS_DEFAULT_TIMEOUT       TIME_T_TO_CDTIME_T (0.015)
```
Comment (mfournier) [inline src/gps.c]: 0.015s is really short, does it really make sense ? What value is there to actually expose this to the user ? How can he adjust this setting to the best possible value for his setup/device ?

File: src/collectd.conf.pod (hunk 16 above)
Comment (NicolasJourden) [inline src/collectd.conf.pod]: Ok. Fixed.

File: src/collectd.conf.pod (hunk 15 above)
Comment (NicolasJourden) [inline src/collectd.conf.pod]: Ok, I will add soon more details.

File: src/collectd.conf.pod (hunk 14 above)
Comment (NicolasJourden) [inline src/collectd.conf.pod]: Fixed.

**Comment by NicolasJourden** [comment]
I simplified a lot the things regarding the pause.
gpsd reading pause was not so useful after watching the traffic.
I have put a range for gpsd timeout in waiting since that is blocking. For me 15 ms works fine, but example in the documentation shows 500 us. The range is now set from 500 us to 5 sec.
Documentation has been updated also.

Let me know if you see anything to improve.
